<a href="https://colab.research.google.com/github/TeraneAgaeva/Data-Analysis-course-/blob/main/%D0%9A%D0%BE%D0%BF%D0%B8%D1%8F_%D0%B1%D0%BB%D0%BE%D0%BA%D0%BD%D0%BE%D1%82%D0%B0_%22hw_3_2_%D0%9E%D1%81%D0%BD%D0%BE%D0%B2%D0%B8_%D1%80%D0%BE%D0%B1%D0%BE%D1%82%D0%B8_%D0%B7_%D0%B4%D0%B0%D0%BD%D0%B8%D0%BC%D0%B8_%D1%83_Pandas_ipynb%22.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Домашнє завдання до теми "Основи роботи з даними у Pandas"

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [1]:
import pandas as pd

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
!ls drive/MyDrive/supermarket_sales.csv

drive/MyDrive/supermarket_sales.csv


In [4]:
data_path = 'drive/MyDrive/supermarket_sales.csv'
df = pd.read_csv(data_path)

1. За допомогою бібліотеки Pandas виведіть перші 4 і останні 4 записи у наборі даних (це дві окремі команди). Треба саме 4! Не 5.

In [5]:
df.head(4)

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
2  631-41-3108      A     Yangon        Normal    Male   
3  123-19-1176      A     Yangon        Member    Male   

             Product line  Unit price  Quantity   Tax 5%     Total       Date  \
0       Health and beauty       74.69         7  26.1415  548.9715   1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200   3/8/2019   
2      Home and lifestyle       46.33         7  16.2155  340.5255   3/3/2019   
3       Health and beauty       58.22         8  23.2880  489.0480  1/27/2019   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  
0  13:08      Ewallet  522.83                 4.761905       26.1415     9.1  
1  10:29         Cash   76.40                 4.761905        3.8200     9.6  
2  13:23  Credit card  324.31                 4.761905       16.2155     7.4  
3  20:33      Ewallet  465.76                 4.761905       23.2880     8.4

In [6]:
df.tail(4)

Invoice ID Branch      City Customer type  Gender         Product line  \
996  303-96-2227      B  Mandalay        Normal  Female   Home and lifestyle   
997  727-02-1313      A    Yangon        Member    Male   Food and beverages   
998  347-56-2442      A    Yangon        Normal    Male   Home and lifestyle   
999  849-09-3807      A    Yangon        Member  Female  Fashion accessories   

     Unit price  Quantity  Tax 5%     Total       Date   Time  Payment  \
996       97.38        10  48.690  1022.490   3/2/2019  17:16  Ewallet   
997       31.84         1   1.592    33.432   2/9/2019  13:22     Cash   
998       65.82         1   3.291    69.111  2/22/2019  15:33     Cash   
999       88.34         7  30.919   649.299  2/18/2019  13:28     Cash   

       cogs  gross margin percentage  gross income  Rating  
996  973.80                 4.761905        48.690     4.4  
997   31.84                 4.761905         1.592     7.7  
998   65.82                 4.761905         3.291     4.1  
999  618.38                 4.761905        30.919     6.6

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1F5XlzBOOTMsedpsYqz6NnXjIPTDe6jpx)

2. Виведіть рядки даних з індексами з 615 по 619 включно.

In [7]:
df[615:620]

Invoice ID Branch       City Customer type  Gender  \
615  509-29-3912      A     Yangon        Member  Female   
616  327-40-9673      B   Mandalay        Member    Male   
617  840-19-2096      C  Naypyitaw        Member    Male   
618  828-46-6863      A     Yangon        Member    Male   
619  641-96-3695      C  Naypyitaw        Member  Female   

               Product line  Unit price  Quantity   Tax 5%     Total  \
615       Sports and travel       38.72         9  17.4240  365.9040   
616       Sports and travel       72.60         6  21.7800  457.3800   
617  Electronic accessories       87.91         5  21.9775  461.5275   
618      Food and beverages       98.53         6  29.5590  620.7390   
619     Fashion accessories       43.46         6  13.0380  273.7980   

          Date   Time      Payment    cogs  gross margin percentage  \
615  3/20/2019  12:24      Ewallet  348.48                 4.761905   
616  1/13/2019  19:51         Cash  435.60                 4.761905   
617  3/14/2019  18:10      Ewallet  439.55                 4.761905   
618  1/23/2019  11:22  Credit card  591.18                 4.761905   
619   2/7/2019  17:55      Ewallet  260.76                 4.761905   

     gross income  Rating  
615       17.4240     4.2  
616       21.7800     6.9  
617       21.9775     4.4  
618       29.5590     4.0  
619       13.0380     8.5

In [8]:
df.loc[615:619]

Invoice ID Branch       City Customer type  Gender  \
615  509-29-3912      A     Yangon        Member  Female   
616  327-40-9673      B   Mandalay        Member    Male   
617  840-19-2096      C  Naypyitaw        Member    Male   
618  828-46-6863      A     Yangon        Member    Male   
619  641-96-3695      C  Naypyitaw        Member  Female   

               Product line  Unit price  Quantity   Tax 5%     Total  \
615       Sports and travel       38.72         9  17.4240  365.9040   
616       Sports and travel       72.60         6  21.7800  457.3800   
617  Electronic accessories       87.91         5  21.9775  461.5275   
618      Food and beverages       98.53         6  29.5590  620.7390   
619     Fashion accessories       43.46         6  13.0380  273.7980   

          Date   Time      Payment    cogs  gross margin percentage  \
615  3/20/2019  12:24      Ewallet  348.48                 4.761905   
616  1/13/2019  19:51         Cash  435.60                 4.761905   
617  3/14/2019  18:10      Ewallet  439.55                 4.761905   
618  1/23/2019  11:22  Credit card  591.18                 4.761905   
619   2/7/2019  17:55      Ewallet  260.76                 4.761905   

     gross income  Rating  
615       17.4240     4.2  
616       21.7800     6.9  
617       21.9775     4.4  
618       29.5590     4.0  
619       13.0380     8.5

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1Otufe_TXRmjZxWigzBiVBLCAVUfEReXM)

3. Виведіть назви колонок у наборі даних разом з їхніми типами. Це можна зробити кількома способами, важливо аби в результаті були назви колонок і відповідні їм типи.

In [9]:
df.dtypes # 1st variant

Invoice ID                  object
Branch                      object
City                        object
Customer type               object
Gender                      object
Product line                object
Unit price                 float64
Quantity                     int64
Tax 5%                     float64
Total                      float64
Date                        object
Time                        object
Payment                     object
cogs                       float64
gross margin percentage    float64
gross income               float64
Rating                     float64
dtype: object

In [10]:
df.info()  # 2st variant

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 17 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Invoice ID               1000 non-null   object 
 1   Branch                   1000 non-null   object 
 2   City                     1000 non-null   object 
 3   Customer type            1000 non-null   object 
 4   Gender                   1000 non-null   object 
 5   Product line             1000 non-null   object 
 6   Unit price               1000 non-null   float64
 7   Quantity                 1000 non-null   int64  
 8   Tax 5%                   1000 non-null   float64
 9   Total                    1000 non-null   float64
 10  Date                     1000 non-null   object 
 11  Time                     1000 non-null   object 
 12  Payment                  1000 non-null   object 
 13  cogs                     1000 non-null   float64
 14  gross margin percentage  

Очікуваний результат (один з можливих форматів):

![](https://drive.google.com/uc?export=view&id=1ASh6-S-_f3X9fTQAPIvnHMFZamTBjNkL)

4. Користуючись методом `DataFrame.describe()` виявіть 1, 25, 50, 75 і 99 персентилі колонки `gross income`. Знайдіть, яке значення відповідає 75отму персентилю цієї колонки і що це означає? Як би Ви його пояснили простими словами колегам? Відповідь напишіть текстом в окремій клітинці типу Markdown.

In [11]:
df['gross income'].describe(percentiles=[0.01, 0.25, 0.5, 0.75, 0.99])

count    1000.000000
mean       15.379369
std        11.708825
min         0.508500
1%          0.913735
25%         5.924875
50%        12.088000
75%        22.445250
99%        45.250750
max        49.650000
Name: gross income, dtype: float64



*   75тому персентилю відповідає значення 22.445250 - це значить, що дані по полю gross income на 75% складаються зі значень до 22.445250. Дані, що є більшими за межами даних 75% поля gross income.




In [5]:
info_75_percentile = df['gross income'].describe(percentiles=[0.75])['75%']
info_str = f"75% персентиль рівний {info_75_percentile:.2f}"
print(info_str)

75% персентиль рівний 22.45


Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1bDa1YFUWDsPm78Lnpy2geQND8M1HEQiT)

5. Виведіть кількість інвойсів за кожним унікальним містом в наборі даних.

  Підказка: інвойс - це один рядок в цьому наборі даних, ми можемо цим скористатись аби швидко отримати відповідь.

In [13]:
df.columns

Index(['Invoice ID', 'Branch', 'City', 'Customer type', 'Gender',
       'Product line', 'Unit price', 'Quantity', 'Tax 5%', 'Total', 'Date',
       'Time', 'Payment', 'cogs', 'gross margin percentage', 'gross income',
       'Rating'],
      dtype='object')

In [14]:
df['City'].value_counts()

City
Yangon       340
Mandalay     332
Naypyitaw    328
Name: count, dtype: int64

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1G8c6M875NOW7oU3OtZegTOnyD2y8lhAV)

6. Виведіть, який відсоток інвойсів припадає на кожне унікальне місто.

In [15]:
df['City'].value_counts(normalize=True)

City
Yangon       0.340
Mandalay     0.332
Naypyitaw    0.328
Name: proportion, dtype: float64

In [16]:
df['City'].value_counts()/df['City'].value_counts().sum()

City
Yangon       0.340
Mandalay     0.332
Naypyitaw    0.328
Name: count, dtype: float64

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1yZZpWhL2y8DX9rSRioYCRh1VbJgcf3fY)

7.1. Використовуючи фільтрацію даних, знайдіть середнє значення `gross income` для міст `Yangon` та `Mandalay`. В якому місті воно більше?

7.2 Сформулюйте 3 гіпотези, чому в одному місті gross income вище, як можна дізнатись чи правдива ця гіпотеза з використанням даних (наявних і додаткових), та як висновки з цих гіпотез можна використати для прийняття рішень як підвищити прибуток в мережі супермаркетів. Відповідь напишіть текстом в окремій клітинці типу Markdown.

Це завдання - на продуктове мислення. Якщо не знаєте, які гіпотези сформулювати - напишіть ті, які здаються вам зовсім примітивними. Важливо навчитись висловлювати свої гіпотези щодо даних, і з часом та досвідом вони будуть ставати щоразу глибше і точніше. Весь аналіз даних будується на тому, що ми перевіряємо певні гіпотези та робимо висновки і рекоемндації на основі своїх досліджень.

In [34]:
mean_gross_income_yangon = df[df['City'] == 'Yangon']['gross income'].mean()
mean_gross_income_mandalay = df[df['City'] == 'Mandalay']['gross income'].mean()
print(mean_gross_income_yangon)
print(mean_gross_income_mandalay)


14.874001470588235
15.232024096385542


Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1JdXUmK46WcoXUJYP7NS2SJlr2XjBbpQ8)

In [35]:
df.head()

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
2  631-41-3108      A     Yangon        Normal    Male   
3  123-19-1176      A     Yangon        Member    Male   
4  373-73-7910      A     Yangon        Normal    Male   

             Product line  Unit price  Quantity   Tax 5%     Total       Date  \
0       Health and beauty       74.69         7  26.1415  548.9715   1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200   3/8/2019   
2      Home and lifestyle       46.33         7  16.2155  340.5255   3/3/2019   
3       Health and beauty       58.22         8  23.2880  489.0480  1/27/2019   
4       Sports and travel       86.31         7  30.2085  634.3785   2/8/2019   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  
0  13:08      Ewallet  522.83                 4.761905       26.1415     9.1  
1  10:29         Cash   76.40                 4.761905        3.8200     9.6  
2  13:23  Credit card  324.31                 4.761905       16.2155     7.4  
3  20:33      Ewallet  465.76                 4.761905       23.2880     8.4  
4  10:37      Ewallet  604.17                 4.761905       30.2085     5.3

In [46]:
df.groupby(['Product line', 'City'])['gross income'].agg(['mean', 'min', 'max']).round(3)

mean    min     max
Product line           City                            
Electronic accessories Mandalay   14.763  1.272  44.878
                       Naypyitaw  16.423  1.512  41.170
                       Yangon     14.537  1.448  44.335
Fashion accessories    Mandalay   12.606  1.676  41.625
                       Naypyitaw  15.795  0.627  49.650
                       Yangon     15.250  0.604  49.490
Food and beverages     Mandalay   14.490  2.686  42.315
                       Naypyitaw  17.148  1.079  49.260
                       Yangon     14.091  1.592  44.397
Health and beauty      Mandalay   17.952  0.888  43.935
                       Naypyitaw  15.216  1.537  45.250
                       Yangon     12.764  0.916  35.840
Home and lifestyle     Mandalay   16.713  1.616  48.690
                       Naypyitaw  14.704  0.699  48.750
                       Yangon     16.423  0.914  45.325
Sports and travel      Mandalay   15.352  1.649  44.982
                       Naypyitaw  16.679  0.508  47.720
                       Yangon     15.636  0.767  44.140

In [45]:
df.groupby(['Product line', 'City'])[['Quantity', 'Unit price']].agg(['mean', 'min', 'max']).round(3)

Quantity         Unit price              
                                     mean min max       mean    min    max
Product line           City                                               
Electronic accessories Mandalay     5.745   1  10     49.854  10.59  99.73
                       Naypyitaw    6.055   1  10     55.809  11.81  99.69
                       Yangon       5.367   1  10     54.871  10.56  99.56
Fashion accessories    Mandalay     4.790   1  10     54.844  14.48  99.89
                       Naypyitaw    5.262   1  10     59.736  10.18  99.82
                       Yangon       5.157   1  10     56.670  12.09  99.10
Food and beverages     Mandalay     5.400   1  10     55.540  17.63  99.69
                       Naypyitaw    5.591   1  10     57.273  14.87  99.79
                       Yangon       5.397   1  10     54.974  10.13  99.78
Health and beauty      Mandalay     6.038   1  10     58.186  10.75  99.71
                       Naypyitaw    5.327   1  10     55.971  10.16  99.96
                       Yangon       5.468   1  10     49.862  10.08  99.83
Home and lifestyle     Mandalay     5.900   1  10     55.514  11.28  99.92
                       Naypyitaw    5.444   1  10     54.334  10.53  97.50
                       Yangon       5.708   1  10     55.846  11.43  97.94
Sports and travel      Mandalay     5.194   1  10     59.678  11.85  99.96
                       Naypyitaw    5.889   1  10     55.107  10.17  99.24
                       Yangon       5.644   1  10     55.610  12.34  98.40

7.2. Відповідь: Можливі причини різниці gross income по 2м містам:

*   Mandalay трохи перевищує середня дохідність gross income над Yangon, адже кількість покупок з товарів першої необхідності було більше здійснено у  Mandalay за категоріями: Food and beverages, Health and beauty, Home and lifestyle.
*   Саме у Mandalay деякі з цих категорій товарів дорожчі і продано більше
*   Більше продано товарів першої необхідності у Mandalay про що може свідчити або більша чисельність населення, або рівень дохідності населення у цьому місті, або туристичний сезон був більш насичений саме у Mandalay






8. Для продовження дослідження різниці у прибутках в магазинах давайте зʼясуємо, чи відрізняється маржа ('gross margin percentage') для різних інвойсів в магазині? Перевірте це програмно на даних виходячи з того функціоналу, який ми вже пройшли у лекціях.

Дайте на основі отриманого результату відповідь: "У всіх всіх інвойсах однакова маржа" або "Маржа відрізняється для різних інвойсів".

In [48]:
df.groupby(['Invoice ID'])['gross margin percentage'].agg(['mean', 'min', 'max']).round(3)


mean    min    max
Invoice ID                      
101-17-6199  4.762  4.762  4.762
101-81-4070  4.762  4.762  4.762
102-06-2002  4.762  4.762  4.762
102-77-2261  4.762  4.762  4.762
105-10-6182  4.762  4.762  4.762
...            ...    ...    ...
894-41-5205  4.762  4.762  4.762
895-03-6665  4.762  4.762  4.762
895-66-0685  4.762  4.762  4.762
896-34-0956  4.762  4.762  4.762
898-04-2717  4.762  4.762  4.762

[1000 rows x 3 columns]

In [49]:
df.groupby(['Product line'])['gross margin percentage'].agg(['mean', 'min', 'max']).round(3)

mean    min    max
Product line                               
Electronic accessories  4.762  4.762  4.762
Fashion accessories     4.762  4.762  4.762
Food and beverages      4.762  4.762  4.762
Health and beauty       4.762  4.762  4.762
Home and lifestyle      4.762  4.762  4.762
Sports and travel       4.762  4.762  4.762

Відповідь:gross margin percentage не відрізняється за інвойсами та групами товарів

Очікуваний результат: відповідь щодо маржі в усіх інвойсах на основі отриманих результатів

9. Врахуовуючи висновок в попередньому завданні, дослідимо як продаються в цих містах різні категорії товарів. Виведіть відсотковий розпоіділ кількості інвойсів по категоріям товарів `Product line` в містах `Yangon` та `Mandalay`.

In [8]:
group1 = df.groupby(['Product line', 'City'])[['Invoice ID']].count()
group1

Invoice ID
Product line           City                 
Electronic accessories Mandalay           55
                       Naypyitaw          55
                       Yangon             60
Fashion accessories    Mandalay           62
                       Naypyitaw          65
                       Yangon             51
Food and beverages     Mandalay           50
                       Naypyitaw          66
                       Yangon             58
Health and beauty      Mandalay           53
                       Naypyitaw          52
                       Yangon             47
Home and lifestyle     Mandalay           50
                       Naypyitaw          45
                       Yangon             65
Sports and travel      Mandalay           62
                       Naypyitaw          45
                       Yangon             59

In [64]:
group1.

Invoice ID
Product line           City                 
Electronic accessories Mandalay           55
                       Naypyitaw          55
                       Yangon             60
Fashion accessories    Mandalay           62
                       Naypyitaw          65
                       Yangon             51
Food and beverages     Mandalay           50
                       Naypyitaw          66
                       Yangon             58
Health and beauty      Mandalay           53
                       Naypyitaw          52
                       Yangon             47
Home and lifestyle     Mandalay           50
                       Naypyitaw          45
                       Yangon             65
Sports and travel      Mandalay           62
                       Naypyitaw          45
                       Yangon             59

In [14]:
filter_1 = (df['City'] == 'Yangon')
invoice_counts = df[filter_1]['Product line'].value_counts(normalize=True)
filter_2 = (df['City'] == 'Mandalay')
invoice_counts2 = df[filter_2]['Product line'].value_counts(normalize=True)

print(invoice_counts)
print(invoice_counts2)

Product line
Home and lifestyle        0.191176
Electronic accessories    0.176471
Sports and travel         0.173529
Food and beverages        0.170588
Fashion accessories       0.150000
Health and beauty         0.138235
Name: proportion, dtype: float64
Product line
Fashion accessories       0.186747
Sports and travel         0.186747
Electronic accessories    0.165663
Health and beauty         0.159639
Food and beverages        0.150602
Home and lifestyle        0.150602
Name: proportion, dtype: float64


Очікуваний результат для кожного з міст:

![](https://drive.google.com/uc?export=view&id=1IbG58YK5KVqC-wCnllF1btgEgfWTbi_O)

![](https://drive.google.com/uc?export=view&id=1aKOkLmlMq1sOh4GbwyaQyjfhNJzdU4Yh)